### ISSUE TITLE
Controlling Acknowledgements explicitly upon receiving messages 

### ISSUE THREAD

**Issue by prabathabey** [mqttjs/MQTT.js#691]
Does MQTT.js client API provide support for controlling the process of sending acknowledgements upon receiving messages explicitly? I'm trying to ensure there is zero message loss in the end-to-end processing path of a request with QoS > 1 and clean_session set to "false". So, whenever I receive a message at a particular message receiver that is subscribed to receive messages published to a particular topic, I need to be sure that the downstream business functionalities are successfully completed before I send PUBACK message back to the broker. Can this be achieved with the functionalities currently available in the MQTT.js client API? Appreciate any feedback on this.

**Comment by mcollina** [comment]
You should use `handleMessage`.

**Comment by prabathabey** [comment]
@mcollina Thanks for the feedback. I looked into the source code and figured out that, in the current implementation, puback is already sent by the time handleMessage is invoked when QoS 1 messages are processed, which does not quite allow me to achieve what's looked for. I would like to suggest a possible solution for this in the form of a fix, which goes below.

```
MqttClient.prototype._handlePublish = function (packet, done) {
     ...
     case 1:
      // emit the message event
      this.emit('message', topic, message, packet)
      this.handleMessage(packet, done)
      if (this.options.autoAcknowledgement === true) {
        this.acknowledge(packet)
      }
      break
     case 0:
      // emit the message event
       this.emit('message', topic, message, packet)
       this.handleMessage(packet, done)
       break
       ...
}

/**
 * Handle explicit acknowledgements in scenarios where the client
 * needs to have more control over signalling the broker of the
 * successful receipt of messages.
 *
 * @param packet  Packet to be acknowledged
 */
MqttClient.prototype.acknowledge = function (packet) {
  var qos = packet.qos
  switch (qos) {
    case 1:
      this._sendPacket({
        cmd: 'puback',
        messageId: packet.messageId
      })
      break
    default:
      // do nothing
      break
  }
}
```

WDYT?

**Comment by mcollina** [comment]
As MQTT 3.1.1 states:

>  The receiver is not required to complete delivery of the Application Message before sending the PUBACK. When its original sender receives the PUBACK packet, ownership of the Application Message is transferred to the receiver.

Instead of introducing a newer API, I would actually prefer it `PUBACK` was sent after `handleMessage` is completed (with an option). Would you mind sending a PR?

**Comment by prabathabey** [comment]
@mcollina Thanks for the feedback. I will surely submit a PR.

**Comment by prabathabey** [comment]
@mcollina I've submitted a PR with the above change, as well as a fix for a test failure I've come across while running the test cases at [1]. I'd appreciate you can review the PR and share your feedback. 

[1] https://github.com/mqttjs/MQTT.js/pull/697

**Comment by prabathabey** [comment]
Closing the ticket as the fix for the issue indicated above is merged into the master branch in PR #697.
